support/k8s/k8s-workload-registrar/mode-crd/controllers/identity_schema.go

```diff
@@ -0,0 +1,313 @@
+		if err != nil {
+			log.Printf("* identity_schema, getFieldValue: Error processing the field %s, with configMap source: %v", field.Name, err)
+			// TODO for now, when error return the field name
+			return field.Name
+		}
+		return value
+	}
+
+	// TODO for now if value unknown, just return the field name
+	log.Printf("* identity_schema, getFieldValue: Error processing the field %s, no matching source!", field.Name)
+	return field.Name
+}
+
+func (att *AttestorSource) GetValue(pod *corev1.Pod) (value string, err error) {
+
+	log.Printf("** identity_schema, GetValue(Attestor): pod: %s", pod.Name)
+	// log.Printf("** This attestor uses mapping: %#v", attestor.Mapping)
+
+	switch att.Group {
+	case nodeAttestor:
+		log.Print("**identity_schema, GetValue: Processing nodeAttestor")
+		return "value-from-node-Attestor", nil
+	case workloadAttestor:
+		// here we don't need the selector values. We just need the value for the SPIFFE ID
+		_, value, err = att.getValueFromWorkloadAttestor(pod)
+		if err != nil {
+			log.Printf("%s", err)
+			return value, err
+		} else {
+			return value, nil
+		}
+	default:
+		log.Print("** identity_schema, GetValue: Unknown attestor name")
+		err := fmt.Errorf("Unknown attestor name: %s", att.Group)
+		return value, err
+	}
+}
+
+func (cms *ConfigMapSource) GetValue(pod *corev1.Pod) (value string, err error) {
+
+	log.Printf("** identity_schema, GetValue(ConfigMap): ConfigMap namespace: %s, name: %s, field: %s", cms.Namespace, cms.Name, cms.Field)
+	err = fmt.Errorf("Function not implemented for %s", "configMapSource")
+	return value, err
+}
+
+// getValueFromWorkloadAttestor is a function used by both getSVID and getSelector
+// in case of the getSVID, the selectorName should be ignored
+// in case of the getSelector, values with empty selectorNames should be ignored
+func (att *AttestorSource) getValueFromWorkloadAttestor(pod *corev1.Pod) (selectorName string, fieldValue string, err error) {
+
+	log.Printf("*** identity_schema, getValueWorkloadAttestor for pod %s", pod.Name)
+	for _, field := range att.Mapping {
+
+		// only certain fields are valid as selectors, other will be ignored
+		switch field.Type {
+		case "k8s":
+			switch field.Field {
+			case "sa":
+				return serviceAccountLabel, pod.Spec.ServiceAccountName, nil
+			case "ns":
+				return namespaceLabel, pod.Namespace, nil
+			case "pod-name":
+				return podNameLabel, pod.Name, nil
+			case "pod-uid":
+				return podUIDLabel, string(pod.UID), nil
+			default:
+				err := fmt.Errorf("Unknown field for k8s attestor: %s", field.Field)
+				log.Printf("%s", err)
+			}
+		// to be used by other attestor types
+		case "xxx":
+			log.Printf("*** Processing xxx attestor")
+		default:
+			err := fmt.Errorf("Unknown attestor type: %s", field.Type)
+			log.Printf("%s", err)
+		}
+	}
+	return selectorName, fieldValue, fmt.Errorf("Cannot find a mapping match. Error: %s", err)
+}
+
+func (is *IdentitySchema) getSelector(pod *corev1.Pod) spiffeidv1beta1.Selector {
```

Comment:
i think it probably would be easier to have the interface that gets the value return some indication of selectors that are required, then just one function that takes in a map of selectors and creates it from there instead of trying to run it all again from scratch. This may create a race resulting in a different selector vs spiffe ID for the same field (getting config map twice, 1 second apart)

Reply:
One method of doing this can be `SelectorOptions` that can be defined like `func (*spiffeidv1beta1.Selector)`, similar to containerd options, so the set of functions will be used to end up resolving what the final selector object looks like by applying all the different options to it. i.e. https://github.com/containerd/containerd/blob/261c107ffc4ff681bc73988f64e3f60c32233b37/container_opts.go